velox/functions/prestosql/aggregates/PrestoHasher.cpp

```diff
@@ -0,0 +1,269 @@
+    BufferPtr& buffer) {
+  bulkHashIntegral<int32_t>(*vector_.get(), rows, buffer);
+}
+
+template <>
+void PrestoHasher::bulkHash<TypeKind::SMALLINT>(
+    const SelectivityVector& rows,
+    BufferPtr& buffer) {
+  bulkHashIntegral<int16_t>(*vector_.get(), rows, buffer);
+}
+
+template <>
+void PrestoHasher::bulkHash<TypeKind::TINYINT>(
+    const SelectivityVector& rows,
+    BufferPtr& buffer) {
+  bulkHashIntegral<int8_t>(*vector_.get(), rows, buffer);
+}
+
+template <>
+void PrestoHasher::bulkHash<TypeKind::DOUBLE>(
+    const SelectivityVector& rows,
+    BufferPtr& buffer) {
+  bulkHashIntegral<int64_t>(*vector_.get(), rows, buffer);
+}
+
+template <>
+void PrestoHasher::bulkHash<TypeKind::DATE>(
+    const SelectivityVector& rows,
+    BufferPtr& buffer) {
+  bulkHashFn(rows, buffer, [&](auto row) -> int64_t {
+    return hashInteger(vector_->valueAt<Date>(row).days());
+  });
+}
+
+template <>
+void PrestoHasher::bulkHash<TypeKind::VARCHAR>(
+    const SelectivityVector& rows,
+    BufferPtr& buffer) {
+  bulkHashFn(rows, buffer, [&](auto row) -> int64_t {
+    return hashStringView(*vector_.get(), row);
+  });
+}
+
+template <>
+void PrestoHasher::bulkHash<TypeKind::VARBINARY>(
+    const SelectivityVector& rows,
+    BufferPtr& buffer) {
+  bulkHashFn(rows, buffer, [&](auto row) -> int64_t {
+    return hashStringView(*vector_.get(), row);
+  });
+}
+
+template <>
+void PrestoHasher::bulkHash<TypeKind::REAL>(
+    const SelectivityVector& rows,
+    BufferPtr& buffer) {
+  bulkHashIntegral<int32_t>(*vector_.get(), rows, buffer);
+}
+
+template <>
+void PrestoHasher::bulkHash<TypeKind::TIMESTAMP>(
+    const SelectivityVector& rows,
+    BufferPtr& buffer) {
+  bulkHashFn(rows, buffer, [&](auto row) -> int64_t {
+    return hashInteger((vector_->valueAt<Timestamp>(row)).toMillis());
+  });
+}
+
+template <>
+void PrestoHasher::bulkHash<TypeKind::BOOLEAN>(
+    const SelectivityVector& rows,
+    BufferPtr& buffer) {
+  bulkHashFn(rows, buffer, [&](auto row) -> int64_t {
+    return vector_->valueAt<bool>(row) ? 1231 : 1237;
+  });
+}
+
+template <>
+void PrestoHasher::bulkHash<TypeKind::ARRAY>(
+    const SelectivityVector& rows,
+    BufferPtr& buffer) {
+  auto baseArray = vector_->base()->as<velox::ArrayVector>();
+
+  auto convertedRows = velox::functions::toElementRows(
+      baseArray->elements()->size(), rows, baseArray, vector_->indices());
+
+  BufferPtr childBuffer =
+      AlignedBuffer::allocate<int64_t>(convertedRows.end(), baseArray->pool());
+
+  children_[0]->bulkHash(convertedRows, childBuffer);
+  auto indices = vector_->indices();
+  auto rawSizes = baseArray->rawSizes();
+  auto rawOffsets = baseArray->rawOffsets();
+  auto rawNulls = baseArray->rawNulls();
+  auto rawChildHashes = childBuffer->as<int64_t>();
+  auto hashBuffer = buffer->asMutable<int64_t>();
+
+  rows.applyToSelected([&](auto row) {
+    int64_t hash = 0;
+    if (!(rawNulls && bits::isBitNull(rawNulls, row))) {
```

Comment:
row -> indices[row] ? 

It would be nice to add test coverage to catch this type of bugs.